documentation/book/con-the-upgrade-process.adoc

```diff
@@ -0,0 +1,56 @@
+// This module is included in the following assemblies:
+//
+// assembly-upgrading-kafka-versions.adoc
+
+[id='con-the-upgrade-process-{context}']
+= The upgrade process
+
+Depending on the difference between the version being upgraded from and to, the upgrade process works in one of two ways:
+
+* A three phase update. This is typically used for upgrading between Kafka versions differing in the major or minor versions, for example 2.0.0 to 2.1.0.
+* A single phase update. This is typically used for upgrading between Kafka versions differing in the micro versions, for example 2.0.0 to 2.0.1.
+
+== Three phase update
+
+When the versions of Kafka use a different interbroker protocol version, or log message format version a three phase update is used.
+
+Phase 1::
+The existing cluster is reconfigured so that the brokers will communicate using the old version of the protocol and using the old message format.
+A rolling restart is necessary so that the brokers are using the new image.
+
+Phase 2::
+All the brokers in the cluster are now running the new image.
+While they all _understand_ the new protocol version, they're all _using_ the old version. 
+Similarly they are using the old message format.
+A rolling restart is necessary for them all to start using the new protocol.
+
+Phase 3::
+You can now update clients, one-by-one, to the new version of the Kafka client libraries.
++
+Once all the clients applications are using the new Kafka client libraries the cluster can be reconfigured so that the brokers use the new message format (using `log.message.format.version`).
+A final rolling restart is necessary.
+
+NOTE: It is possible to revert the version change until the point that `log.message.format.version` is updated in step 3.
+Once producers have sent messages in the new message format and those messages have been appended to broker logs it is not possible to revert to the old version because the old brokers won't be able to read the messages in the log.
+
+User action is required as follows:
+
+. Initially to set `log.message.format.version` in the `Kafka.spec.kafka.config` to the message format version is use by the version of Kafka being upgraded _from_. 
+This step is not necessary if `log.message.format.version` already has the correct value.
+
+. To change the `Kafka.spec.kafka.version` to the version being upgraded _to_. 
+The Cluster Operator manages the rolling restarts necessary for Phase 1 and 2.
+
+. In Phase 3, to upgrade the clients and update the `Kafka` resource once that has been done.
+Once that has been done the `log.message.format.version` in the `Kafka.spec.kafka.config` can be set to the default value for the `Kafka.spec.kafka.version` that has been upgraded _to_.
```

Comment:
While the CO can manage the change in the protocol version number (thus the user does not need to set it at all) that's not the case with the message format version. That's because the user needs to upgrade their client applications before switching to the new version. Since the `log.message.format.version` is therefore the mechanism by which the user indicates the progression (or rollback) of the upgrades I think it's better if the CO doesn't try to do anything clever with it and requires that the user sets it explicitly.

Reply:
but when you upgrade to the version you want, if you don't set the `log.message.format.version`, is Kafka going to use the message format for the new upgraded version right? What's the clever thing that CO should do? I am just "worried" even if it's not a big deal that the `Kafka` YAML will always have this `log.message.format.version` property in the `config` section even if it's not needed because the Kafka cluster is using the proper message format for the currently used version.

Reply:
> but when you upgrade to the version you want, if you don't set the log.message.format.version, is Kafka going to use the message format for the new upgraded version right? 

Once all the brokers are using the new version then of course you can omit the `log.message.format.version` to get the default version for the running version of Kafka. If you omitted it prior to upgrading the broker versions then it would be ambiguous. What version is any particular broker using? It's to avoid this ambiguity that the upgrade logic forces the user to specify the `log.message.format.version` explicitly up front, before any version changes happen. Thus the user is in control (and is also responsible) for exactly when `log.message.format.version` gets upgraded to the new version (which is the part which is not reversible).

> What's the clever thing that CO should do?

I don't think there is any clever thing the CO _could_ do because:

1. It doesn't know when the user has upgraded all the client applications.
2. It doesn't know when the user to prepared to commit to the new version.

> I am just "worried" even if it's not a big deal that the Kafka YAML will always have this log.message.format.version property in the config section even if it's not needed because the Kafka cluster is using the proper message format for the currently used version.

TBH I'm not sure I'm parsing that correctly. I thought you meant

> I am just "worried" (even if it's not a big deal that the Kafka YAML will always have this log.message.format.version property in the config section even if it's not needed) because the Kafka cluster is using the proper message format for the currently used version.

but then I don't understand what there is to worry about, so I assume you meant something else.

Reply:
Finally we got the point offline :-)